common/event_emitter.go

```diff
@@ -180,6 +187,17 @@ func (e *BaseEventEmitter) emit(event string, data interface{}) {
 				handlers = append(handlers[:i], handlers[i+1:]...)
 				continue
 			default:
+				// To ensure that goroutines don't break the ordering
+				// of the emitted events, we will need the goroutine to synchronize.
+				// This means that the events need to be stored in a queue per handler. The
+				// goroutine responsible for the the first popped element must complete
+				// publishing the event before the next goroutine can pop and work with
+				// the next event for that one handler. Each handler can process events concurrently still.
+
+				handler.queueMutex.Lock()
+				handler.queue = append(handler.queue, Event{typ: event, data: data})
+				handler.queueMutex.Unlock()
+
 				go emitEvent(handler)
```

Comment:
Since we have one queue that works on a single event at a time per handler, there's no need to spawn a new goroutine per event. Maybe a goroutine per handler? I'll explore this too.

Reply:
Actually, I think it's easier to leave it so that a new goroutine is spawned as it signifies that an event has been received and we need to now send it to the handler. otherwise we might need to implement out own custom channel the encapsulates this behaviour.